// melt/dist/utils/use-floating.svelte.js
import { arrow, offset, flip, shift, computePosition, autoUpdate } from "@floating-ui/dom";
import { isHtmlElement } from "./is";
import { deepMerge } from "./merge";
import { extract } from "./extract";
const ARROW_TRANSFORM = {
    bottom: "rotate(45deg)",
    left: "rotate(135deg)",
    top: "rotate(225deg)",
    right: "rotate(315deg)",
};
export function useFloating(_node, _floating, _options) {
    const nodeEl = $derived(extract(_node));
    const floatingEl = $derived(extract(_floating));
    const options = $derived(extract(_options, {}));
    const compute = () => {
        const arrowEl = floatingEl.querySelector("[data-arrow]");
        const arrowOffset = isHtmlElement(arrowEl) ? arrowEl.offsetHeight / 2 : 0;
        const arrowMiddleware = isHtmlElement(arrowEl) ? arrow({ element: arrowEl }) : undefined;
        const baseOptions = {
            middleware: [shift(), flip(), arrowMiddleware, offset({ mainAxis: 8 + arrowOffset })],
        };
        computePosition(nodeEl, floatingEl, deepMerge(baseOptions, options)).then(({ x, y, placement, middlewareData, strategy }) => {
            const transformOriginMap = {
                top: "bottom center",
                "top-start": "bottom left",
                "top-end": "bottom right",
                bottom: "top center",
                "bottom-start": "top left",
                "bottom-end": "top right",
                left: "center center",
                "left-start": "top left",
                "left-end": "bottom left",
                right: "center center",
                "right-start": "top right",
                "right-end": "bottom right",
            };
            Object.assign(floatingEl.style, {
                position: strategy,
                left: `${x}px`,
                top: `${y}px`,
            });
            const [side, align = "center"] = placement.split("-");
            floatingEl.style.transformOrigin = transformOriginMap[placement];
            floatingEl.dataset.side = side;
            floatingEl.dataset.align = align;
            if (isHtmlElement(arrowEl) && middlewareData.arrow) {
                const { x, y } = middlewareData.arrow;
                const dir = placement.split("-")[0];
                Object.assign(arrowEl.style, {
                    position: "absolute",
                    left: x ? `${x}px` : "",
                    top: y ? `${y}px` : "",
                    [dir]: `calc(100% - ${arrowOffset}px)`,
                    transform: ARROW_TRANSFORM[dir],
                    backgroundColor: "inherit",
                    zIndex: "inherit",
                });
                arrowEl.dataset.side = dir;
            }
        });
    };
    $effect(() => {
        return autoUpdate(nodeEl, floatingEl, compute);
    });
}
